network/http/source/main.c

```diff
@@ -5,41 +5,133 @@
 
 #include <3ds.h>
 
-Result http_download(httpcContext *context)//This error handling needs updated with proper text printing once ctrulib itself supports that.
+Result http_download(char *url)
 {
 	Result ret=0;
+	httpcContext context;
+	char *newurl=NULL;
 	u8* framebuf_top;
 	u32 statuscode=0;
-	u32 size=0, contentsize=0;
-	u8 *buf;
+	u32 contentsize=0, readsize=0, size=0;
+	u8 *buf, *lastbuf;
 
-	ret = httpcBeginRequest(context);
-	if(ret!=0)return ret;
+	printf("Downloading %s\n",url);
+	gfxFlushBuffers();
 
-	ret = httpcGetResponseStatusCode(context, &statuscode, 0);
-	if(ret!=0)return ret;
+	do {
+		ret = httpcOpenContext(&context, HTTPC_METHOD_GET, url, 1);
+		printf("return from httpcOpenContext: %"PRId32"\n",ret);
+		gfxFlushBuffers();
 
-	if(statuscode!=200)return -2;
+		// This disables SSL cert verification, so https:// will be usable
+		ret = httpcSetSSLOpt(&context, SSLCOPT_DisableVerify);
+		printf("return from httpcSetSSLOpt: %"PRId32"\n",ret);
+		gfxFlushBuffers();
 
-	ret=httpcGetDownloadSizeState(context, NULL, &contentsize);
-	if(ret!=0)return ret;
+		// Enable Keep-Alive connections (on by default, pending ctrulib merge)
+		// ret = httpcSetKeepAlive(&context, HTTPC_KEEPALIVE_ENABLED);
+		// printf("return from httpcSetKeepAlive: %"PRId32"\n",ret);
+		// gfxFlushBuffers();
 
-	printf("size: %"PRId32"\n",contentsize);
-	gfxFlushBuffers();
+		// Set a User-Agent header so websites can identify your application
+		ret = httpcAddRequestHeaderField(&context, (char*)"User-Agent", (char*)"httpc-example/1.0.0");
+		printf("return from httpcAddRequestHeaderField: %"PRId32"\n",ret);
+		gfxFlushBuffers();
 
-	buf = (u8*)malloc(contentsize);
-	if(buf==NULL)return -1;
-	memset(buf, 0, contentsize);
+		// Tell the server we can support Keep-Alive connections.
+		// This will delay connection teardown momentarily (typically 5s)
+		// in case there is another request made to the same server.
+		ret = httpcAddRequestHeaderField(&context, (char*)"Connection", (char*)"Keep-Alive");
+		printf("return from httpcAddRequestHeaderField: %"PRId32"\n",ret);
+		gfxFlushBuffers();
 
+		ret = httpcBeginRequest(&context);
+		if(ret!=0){
+			httpcCloseContext(&context);
+			return ret;
+		}
+
+		ret = httpcGetResponseStatusCode(&context, &statuscode, 0);
+		if(ret!=0){
+			httpcCloseContext(&context);
+			return ret;
+		}
+
+		if ((statuscode >= 301 && statuscode <= 303) || (statuscode >= 307 && statuscode <= 308)) {
+			newurl = malloc(0x1000); // One 4K page for new URL
```

Comment:
If the example were a .cpp file, then yes it should be cast. But it's not.
The example with the warning has had it renamed and compiled with cpp. In C
casting the malloc() itself away from void* is commonly considered to be a
bad idea.
On Wed, Jul 27, 2016 at 9:28 AM fincs <notifications@github.com> wrote:

> In network/http/source/main.c
> <https://github.com/devkitPro/3ds-examples/pull/9#discussion_r72473550>:
>
> >
> > +		ret = httpcBeginRequest(&context);
> > +		if(ret!=0){
> > +			httpcCloseContext(&context);
> > +			return ret;
> > +		}
> > +
> > +		ret = httpcGetResponseStatusCode(&context, &statuscode, 0);
> > +		if(ret!=0){
> > +			httpcCloseContext(&context);
> > +			return ret;
> > +		}
> > +
> > +		if ((statuscode >= 301 && statuscode <= 303) || (statuscode >= 307 && statuscode <= 308)) {
> > +			newurl = malloc(0x1000); // One 4K page for new URL
>
> I disagree.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/devkitPro/3ds-examples/pull/9/files/95d43fd511a504523cc43e242a9ca0ae875f29e1#r72473550>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AECriLlXRia__jV96KLG5dw9Tb3DOvS3ks5qZ4cwgaJpZM4JVyxE>
> .
>